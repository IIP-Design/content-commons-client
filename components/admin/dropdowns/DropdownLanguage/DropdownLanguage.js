import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from 'semantic-ui-react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

const DropdownLanguage = props => {
  const { data, ...rest } = props;
  if ( data.loading ) return 'Loading language...';
  if ( data.error ) return 'Loading error...';
  return <Dropdown { ...rest } options={ data.languages } selection />;
};

DropdownLanguage.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.string,
  data: PropTypes.object,
  fluid: PropTypes.bool,
  required: PropTypes.bool
};

const LANGUAGES_QUERY = gql`
  query Languages($orderBy: LanguageOrderByInput) {
    languages(orderBy: $orderBy) {
      value: id
      text: displayName
    }
  }
`;

export default graphql( LANGUAGES_QUERY, {
  options: {
    variables: { orderBy: 'displayName_ASC' }
  }
} )( DropdownLanguage );

export { LANGUAGES_QUERY };
