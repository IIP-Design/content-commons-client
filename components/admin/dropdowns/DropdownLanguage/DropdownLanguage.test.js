import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import DropdownLanguage, { LANGUAGES_QUERY } from './DropdownLanguage';

const props = {
  id: 'm123',
  name: 'fileLanguageId',
  onChange: jest.fn(),
  value: 'eng13',
  fluid: true,
  required: true
};

const mocks = [
  {
    request: {
      query: LANGUAGES_QUERY,
      variables: { orderBy: 'displayName_ASC' }
    },
    result: {
      data: {
        languages: [
          { value: 'eng13', text: 'English' },
          { value: 'fr143', text: 'French' }
        ]
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <DropdownLanguage { ...props } />
  </MockedProvider>
);

describe( '<DropdownLanguage />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const dropdown = wrapper.find( 'DropdownLanguage' );

    expect( dropdown.exists() ).toEqual( true );
    expect( dropdown.contains( 'Loading language...' ) ).toEqual( true );
  } );

  it( 'renders error message & icon if error is thrown', async () => {
    const errorMocks = [
      {
        request: {
          query: LANGUAGES_QUERY,
          variables: { orderBy: 'displayName_ASC' }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];
    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <DropdownLanguage { ...props } />
      </MockedProvider>
    );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const dropdown = wrapper.find( 'DropdownLanguage' );

    expect( dropdown.contains( 'Loading error...' ) ).toEqual( true );
  } );

  it( 'renders the dropdown with correct languages', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const dropdownLanguage = wrapper.find( 'DropdownLanguage' );
    const dropdownSemanticUI = dropdownLanguage.find( 'Dropdown' );
    const renderedLangItems = dropdownSemanticUI.find( 'DropdownItem' );
    const { languages } = dropdownLanguage.prop( 'data' );

    expect( dropdownSemanticUI.exists() ).toEqual( true );
    expect( renderedLangItems.length ).toEqual( languages.length );
    renderedLangItems.forEach( ( item, i ) => {
      expect( item.prop( 'value' ) ).toEqual( languages[i].value );
      expect( item.prop( 'text' ) ).toEqual( languages[i].text );
    } );
    expect( toJSON( dropdownLanguage ) ).toMatchSnapshot();
  } );

  it( 'making a selection calls `onChange`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const dropdown = wrapper.find( 'DropdownLanguage' );

    dropdown.simulate( 'change' );
    expect( props.onChange ).toHaveBeenCalled();
  } );
} );
