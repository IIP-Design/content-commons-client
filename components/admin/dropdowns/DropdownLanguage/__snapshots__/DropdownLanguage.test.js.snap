// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownLanguage /> renders the dropdown with correct languages 1`] = `
<DropdownLanguage
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "languages": Array [
        Object {
          "text": "English",
          "value": "eng13",
        },
        Object {
          "text": "French",
          "value": "fr143",
        },
      ],
      "loading": false,
      "networkStatus": 7,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "orderBy": "displayName_ASC",
      },
    }
  }
  fluid={true}
  id="m123"
  name="fileLanguageId"
  onChange={[MockFunction]}
  required={true}
  value="eng13"
>
  <Dropdown
    additionLabel="Add "
    additionPosition="top"
    closeOnBlur={true}
    deburr={false}
    fluid={true}
    icon="dropdown"
    id="m123"
    minCharacters={1}
    name="fileLanguageId"
    noResultsMessage="No results found."
    onChange={[MockFunction]}
    openOnFocus={true}
    options={
      Array [
        Object {
          "text": "English",
          "value": "eng13",
        },
        Object {
          "text": "French",
          "value": "fr143",
        },
      ]
    }
    renderLabel={[Function]}
    required={true}
    searchInput="text"
    selectOnBlur={true}
    selectOnNavigation={true}
    selection={true}
    value="eng13"
    wrapSelection={true}
  >
    <div
      aria-expanded={false}
      className="ui fluid selection dropdown"
      id="m123"
      name="fileLanguageId"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      required={true}
      role="listbox"
      tabIndex={0}
    >
      <div
        aria-live="polite"
        className="text"
        role="alert"
      >
        English
      </div>
      <Icon
        as="i"
        className=""
        name="dropdown"
        onClick={[Function]}
      >
        <i
          aria-hidden="true"
          className="dropdown icon"
          onClick={[Function]}
        />
      </Icon>
      <DropdownMenu>
        <div
          className="menu transition"
        >
          <DropdownItem
            active={true}
            key="eng13"
            onClick={[Function]}
            selected={true}
            style={
              Object {
                "pointerEvents": "all",
              }
            }
            text="English"
            value="eng13"
          >
            <div
              aria-checked={true}
              aria-selected={true}
              className="active selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                English
              </span>
            </div>
          </DropdownItem>
          <DropdownItem
            active={false}
            key="fr143"
            onClick={[Function]}
            selected={false}
            style={
              Object {
                "pointerEvents": "all",
              }
            }
            text="French"
            value="fr143"
          >
            <div
              aria-checked={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                French
              </span>
            </div>
          </DropdownItem>
        </div>
      </DropdownMenu>
    </div>
  </Dropdown>
</DropdownLanguage>
`;
